torch/csrc/inductor/static_cuda_launcher.cpp

```diff
@@ -0,0 +1,489 @@
+  use a statically compiled library to launch the kernel. To avoid mallocing
+  for the arguments, we have a launcher for different numbers of arguments up
+  to a max. StaticCudaLauncher only supports # of arguments up until 10 for
+  now.
+
+  Note that we allocate 8 bytes per argument, no matter the types of each
+  argument, since we don't know ahead of time what the types of each argument
+  passed to the triton kernel are. This may take slightly more memory on the
+  stack, and will require some benchmarking. However, since the vast majority
+  of triton kernels have less than 10 args, this seems unlikely to be
+  expensive.
+
+  This launcher is paired with StaticallyLaunchedCudaKernel in
+  triton_heuristics.py.
+
+  TODO:
+  - Handle CutensorMap, NvtmDesc
+  - Handle launch_enter and launch_exit hooks (in python maybe?)
+ */
+
+// Use ATen/NVRTC.h to gain access to the CUDA driver API.
+// This function is only called when CUDA is enabled, and only called to load
+// and launch triton compiled CUDA kernels, so CUDA should always be
+// initialized.
+namespace {
+const at::cuda::NVRTC& nvrtc() {
+  return at::globalContext().getNVRTC();
+}
+
+CUdeviceptr getPointer(PyObject* obj) {
+  CUdeviceptr data_ptr = 0;
+  if (THPUtils_checkLong(obj)) {
+    data_ptr = THPUtils_unpackUInt64(obj);
+    return data_ptr;
+  }
+  if (obj == Py_None) {
+    // valid nullptr
+    return data_ptr;
+  }
+  auto ptr = THPObjectPtr{PyObject_GetAttrString(obj, "data_ptr")};
+  if (ptr) {
+    auto empty_tuple = THPObjectPtr{PyTuple_New(0)};
+    auto ret = THPObjectPtr{PyObject_Call(ptr, empty_tuple, nullptr)};
+    if (!THPUtils_checkLong(ret)) {
+      throw std::runtime_error(
+          "data_ptr method of Pointer object must return 64-bit int");
+    }
+    data_ptr = THPUtils_unpackUInt64(ret);
+    if (!data_ptr)
+      return data_ptr;
+
+    CUdeviceptr dev_ptr = 0;
+    AT_CUDA_DRIVER_CHECK(nvrtc().cuPointerGetAttribute(
+        &dev_ptr, CU_POINTER_ATTRIBUTE_DEVICE_POINTER, data_ptr));
+    return dev_ptr;
+  }
+  throw std::runtime_error(
+      "Pointer argument must be either uint64 or have data_ptr method");
+}
+
+CUfunction loadKernel(
+    std::string filePath,
+    const std::string& funcName,
+    uint32_t sharedMemBytes,
+    const std::optional<std::string>& cubinDir = std::nullopt) {
+  if (cubinDir) {
+    std::filesystem::path p1{*cubinDir};
+    std::filesystem::path p2{filePath};
+    filePath = (p1 / p2.filename()).string();
+  }
+
+  CUmodule mod = nullptr;
+  CUfunction func = nullptr;
+  AT_CUDA_DRIVER_CHECK(nvrtc().cuModuleLoad(&mod, filePath.c_str()));
+  AT_CUDA_DRIVER_CHECK(
+      nvrtc().cuModuleGetFunction(&func, mod, funcName.c_str()));
+  if (sharedMemBytes > 0) {
+    AT_CUDA_DRIVER_CHECK(nvrtc().cuFuncSetAttribute(
+        func, CU_FUNC_ATTRIBUTE_MAX_DYNAMIC_SHARED_SIZE_BYTES, sharedMemBytes));
+  }
+  return func;
+}
+
+template <size_t NUM_ARGS>
```

Comment:
why is this function templated? You don't need to pass std::array args, you only need args.data()